hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -203,14 +197,14 @@ public HoodieData<HoodieRecord<HoodieMetadataPayload>> getRecordsByKeyPrefixes(L
 
               return mergedRecords.stream()
                 .map(keyRecordPair -> keyRecordPair.getValue().orElse(null))
+                .filter(Objects::nonNull)
                 .iterator();
             } catch (IOException ioe) {
               throw new HoodieIOException("Error merging records from metadata table for  " + sortedKeyPrefixes.size() + " key : ", ioe);
             } finally {
               closeReader(readers);
             }
-          })
-        .filter(Objects::nonNull);
+          });
   }
 
   @Override
```

Comment:
I see that for getRecordsByKeyPrefixes(), we parallelize among each file size, where as for getRecordsByKeys, we do foreach (do not parallelize). wondering if we should parallelize getRecordsByKeys() calls too ?

Reply:
for colstats, I do see we do parallelize at higher layer, but thinking for other calls if we can try and leverage spark context to parallelize calls across file slics. for files partitions, there is only one file slice. so, might be a no-op. but for bloom

Reply:
Correct. Check my other comment -- much bigger cleanup needed here. Will create a ticket to follow-up

Reply:
There's actually one already: HUDI-5556